mipengine/controller/node_landscape_aggregator.py

```diff
@@ -207,70 +224,114 @@ def stop(self):
-        with self._update_lock:
-            return self._data_model_registry.get_all_available_datasets_per_data_model()
+        return (
+            self._registries.data_model_registry.get_all_available_datasets_per_data_model()
+        )
 
     def data_model_exists(self, data_model: str) -> bool:
-        with self._update_lock:
-            return self._data_model_registry.data_model_exists(data_model)
+        return self._registries.data_model_registry.data_model_exists(data_model)
 
     def dataset_exists(self, data_model: str, dataset: str) -> bool:
-        with self._update_lock:
-            return self._data_model_registry.dataset_exists(data_model, dataset)
+        return self._registries.data_model_registry.dataset_exists(data_model, dataset)
 
     def get_node_ids_with_any_of_datasets(
         self, data_model: str, datasets: List[str]
     ) -> List[str]:
-        with self._update_lock:
-            return self._data_model_registry.get_node_ids_with_any_of_datasets(
-                data_model, datasets
-            )
+        return self._registries.data_model_registry.get_node_ids_with_any_of_datasets(
+            data_model, datasets
+        )
 
     def get_node_specific_datasets(
         self, node_id: str, data_model: str, wanted_datasets: List[str]
     ) -> List[str]:
-        with self._update_lock:
-            return self._data_model_registry.get_node_specific_datasets(
-                node_id, data_model, wanted_datasets
-            )
+        return self._registries.data_model_registry.get_node_specific_datasets(
+            node_id, data_model, wanted_datasets
+        )
+
+
+def _get_datasets_per_node(
+    nodes: Dict[str, NodeInfo],
+) -> Dict[str, Dict[str, Dict[str, str]]]:
+    datasets_per_node = {}
+    for node_id, node_info in nodes.items():
+        node_socket_addr = _get_node_socket_addr(node_info)
+        datasets_per_data_model = _get_node_datasets_per_data_model(node_socket_addr)
+        if datasets_per_data_model:
+            datasets_per_node[node_info.id] = datasets_per_data_model
+
+    return datasets_per_node
+
+
+def remove_duplicated_datasets(datasets_per_node):
+    aggregated_datasets = [
+        dataset
+        for datasets_per_data_model in datasets_per_node.values()
+        for datasets in datasets_per_data_model.values()
+        for dataset in datasets
+    ]
+    duplicated = [
+        item for item, count in Counter(aggregated_datasets).items() if count > 1
+    ]
+    for node, datasets_per_data_model in datasets_per_node.items():
+        for data_model, datasets in datasets_per_data_model.items():
+            datasets_per_node[node][data_model] = {
+                dataset_name: dataset_label
+                for dataset_name, dataset_label in datasets.items()
+                if dataset_name not in duplicated
+            }
 
 
 def _gather_all_dataset_info(
-    nodes: List[NodeInfo],
+    datasets_per_node: Dict[str, Dict[str, Dict[str, str]]],
```

Comment:
Deeply nested data structures make it difficult to understand what kind of information they re holding. It is better to define a DTO class, where there are named properties which make much more obvious what the intention is